### PR TITLE
feat(examples-plugins): add knip to example plugin

### PR DESCRIPTION
**Pull request by BioPhoton**
This PR is just here to showcase `knip` implementation and screenshots.

**This PR includes:**
- examples-plugins/src
  - logic to wrap the knip CLI
  - unit and integration tests 
  - docs
- utils
  - helper to setup categories `capital` and `singular`
  
**This PR does not include:**
- e2e tests because it is an example

See the portal here: https://staging.code-pushup.dev/portal/code-pushup/cli-utils/commit/985ac7155f1f8cb8ad1b482a984740ad3d2d5e75/dashboard

**Changes for a followup PR:**  
- issues table - `duplicated exports` should list last reference instead of first
- issues table - `unused dependencies` should link to line where package sit's
- issues table - `unlisted dependencies`, `unused development dependencies` could list only one referencing file (otherwise we see multiple times the same issue title with a difference source but installing it ones fixes the issues)
-  issues table - issue messages could include the full audit title e.g. "Unused export docsUrlSchema" instead of "Export docsUrlSchema"
 

## Used configuration

```ts
import {
  KnipConfigPlugin,
  combineNxKnipPlugins,
  withEsbuildApps,
  withEsbuildPublishableLibs,
  withEslint,
  withLibraryMapper,
  withLocalNxPlugins,
  withNxTsPaths,
  withVitest,
} from '@beaussan/nx-knip';

export const withIgnoreMockInLibs = () =>
  withLibraryMapper({
    mapperFn: ({ rootFolder }) => {
      return {
        ignore: [rootFolder + '/mocks/**'],
        entry: [rootFolder + '/src/bin.ts', rootFolder + '/perf/**/index.ts'],
      };
    },
  });

export const withCuNxStandards = (): KnipConfigPlugin => () => {
  return {
    project: ['**/*.{ts,js,tsx,jsx}'],
    ignore: ['tmp/**', 'node_modules/**', 'examples/**'],
    entry: [
      'testing/test-utils/src/index.ts',
      'testing/test-utils/src/lib/fixtures/configs/*.{js,mjs,ts,cjs,mts,cts}',
      'testing/test-setup/src/index.ts',
      'testing/test-setup/src/lib/**/*.{js,mjs,ts,cjs,mts,cts}',
      // Same issue as the other vitest related, it should be picked up by knip from the vitest setup files
      'global-setup.ts',
      'global-setup.e2e.ts',
      // missing knip plugin for now, so this is in the root entry
      'packages/models/zod2md.config.ts',
      'code-pushup.config.ts',
      'esbuild.config.js',
      'tools/**/*.{js,mjs,ts,cjs,mts,cts}',
    ],
    ignoreDependencies: [
      'prettier',
      '@swc/helpers',
      '@swc/cli',
      '@nx/plugin',
      '@nx/workspace',
      '@example/custom-plugin',
    ],
  };
};

export default combineNxKnipPlugins(
  withNxTsPaths(),
  withLocalNxPlugins({ pluginNames: ['nx-plugin'] }),
  withEsbuildApps(),
  withEsbuildPublishableLibs(),
  withVitest(),
  withIgnoreMockInLibs(),
  withEslint(),
  withCuNxStandards(),
);
```

run `nx run-collect --onlyPlugins knip` to test it locally.

## Example Audit Interpretation

**Legend:**  
- ✅ - valid result
- ❌ - false positive
- ❓ - open question
- ⏳ - evaluation pending

**Unused files (1)**  
- ✅ packages/plugin-lighthouse/src/index.ts
Can be fixed by using it in e.g. e2e tests

**Unused dependencies (1)**  
- ✅ cli-table3  package.json
Can be fixed by uninstalling `cli-table3`
Run `npm remove cli-table3` to remove dependency

**Unused devDependencies (8)**  
- ✅  tsx                                    package.json
  This package is not used anywhere in the code either directly on in "scripts" or configuration files. 
  You can fix it by running `npm remove tsx`
  Same goes for the follwing dependencies
- ✅  @trivago/prettier-plugin-sort-imports  package.json
- ⏳ eslint-plugin-jsx-a11y                 package.json
- ❓  eslint-plugin-react                    package.json
should have been picked up by the nx helper WDYT @beaussan
the following packages are (most probably) a cascade of this issue
- ❓  eslint-plugin-react-hooks              package.json
- ❓  jsonc                                  package.json
- ❌ commitlint-plugin-tense                package.json
false as it is used in `commitlint.config.js`

```ts
/** @type {import('@commitlint/types').UserConfig} */
const configuration = {
  extends: ['@commitlint/config-conventional'],
  plugins: ['commitlint-plugin-tense'],
// ...
}
```
- ❌ conventional-changelog-angular         package.json
false as a peer dependency of commit. (cascade of `commitlint-plugin-tense`)
Run `npm ls conventional-changelog-angular` to list check dependent packages.
```bash
├─┬ @commitlint/cli@17.8.1
│ └─┬ @commitlint/lint@17.8.1
│   └─┬ @commitlint/parse@17.8.1
│     └── conventional-changelog-angular@6.0.0
└─┬ commitlint-plugin-tense@1.0.2
  └─┬ @commitlint/lint@19.1.0
    └─┬ @commitlint/parse@19.0.3
      └── conventional-changelog-angular@7.0.0

```

**Unlisted dependencies (38)**  
- ✅ jsonc-eslint-parser                   .eslintrc.json
  This is a peer dependency of `@nx/eslint-plugin`, but and not directly installed. 
  If you uninstall `@nx/eslint-plugin` you don't have `jsonc-eslint-parser` anymore available.  
  Run `npm ls jsonc-eslint-parser` to list the root package:
```bash
@code-pushup/cli-source@0.29.0 /Users/michael_hladky/WebstormProjects/quality-metrics-cli
└─┬ @nx/eslint-plugin@17.3.2
  └── jsonc-eslint-parser@2.4.0
   ```   
   To fix it run `npm i jsonc-eslint-parser --save-dev`
   same is valid for all following issues of `jsonc-eslint-parser`                                                                                                               
- ✅ jsonc-eslint-parser                   e2e/cli-e2e/.eslintrc.json                                                      
- ✅ jsonc-eslint-parser                   packages/cli/.eslintrc.json                                                     
- ✅ jsonc-eslint-parser                   packages/core/.eslintrc.json                                                    
- ✅ jsonc-eslint-parser                   packages/models/.eslintrc.json                                                  
- ✅ jsonc-eslint-parser                   packages/nx-plugin/.eslintrc.json                                               
- ✅ jsonc-eslint-parser                   packages/plugin-coverage/.eslintrc.json                                         
- ✅ jsonc-eslint-parser                   packages/plugin-eslint/.eslintrc.json                                           
- ✅ jsonc-eslint-parser                   packages/plugin-js-packages/.eslintrc.json                                      
- ✅ jsonc-eslint-parser                   packages/plugin-lighthouse/.eslintrc.json                                       
- ✅ jsonc-eslint-parser                   packages/utils/.eslintrc.json                                                   
- ✅ jsonc-eslint-parser                   testing/test-setup/.eslintrc.json                                               
- ✅ jsonc-eslint-parser                   testing/test-utils/.eslintrc.json        
- ✅ minimatch                             tools/eslint-to-code-pushup.mjs                   
This is used in `tools/eslint-to-code-pushup.mjs`, but and not listed in `package.json`. The code works because other packages like `@nx/devkit` have `minimatch` as peer dependency. If you would remove all related packages like `@nx/devkit`, `tools/eslint-to-code-pushup.mjs`would stop working.
  Run `npm ls minimatch` to list the root package:
```bash
├─┬ @nx/devkit@17.3.2
│ └─┬ ejs@3.1.9
│   └─┬ jake@10.8.7
│     ├─┬ filelist@1.0.4
│     │ └── minimatch@5.1.6
│     └── minimatch@3.1.2
 ```
   
   To fix it run `npm i minimatch --save-dev`
   same is valid for issues of `tslib`      
- ✅ tslib                                 tsconfig.base.json     
- ✅ @example/custom-plugin                testing/test-utils/src/lib/fixtures/configs/code-pushup.needs-tsconfig.config.ts
  This is used as dummy reference in a test like this.
  
```ts
  // the point is to test runtime import which relies on alias defined in tsconfig.json "paths"
// non-type import from '@example/custom-plugin' wouldn't work without --tsconfig
import customPlugin from '@example/custom-plugin';

const config = {
  plugins: [customPlugin],
};

export default config;
```
  
It can be fixed by adding the package name to the `knip.config.ts`
```ts
 export default { 
// ...
  ignoreDependencies: [
      // 👇 fixed by add package name to ignoreDependencies here
      '@example/custom-plugin'
    ],
}
```    
- ✅ @commitlint/types                     commitlint.config.js      
Is used in like this: `const { RuleConfigSeverity } = require('@commitlint/types');`.
To fix it run `npm i @commitlint/types --save-dev`
                                                   
- ❓ basic                                 e2e/cli-e2e/vite.config.e2e.ts      
This shows no dependents when running `npm ls basic`  
This could be fixed by installing `basic` manually over `npm i basic -D` but seems odd as there are no deps listed.  
Same goes for all other issues related to `basic`                                              
- basic                                 packages/cli/vite.config.integration.ts                                         
- basic                                 packages/cli/vite.config.unit.ts                                                
- basic                                 packages/core/vite.config.integration.ts                                        
- basic                                 packages/core/vite.config.unit.ts                                               
- basic                                 packages/models/vite.config.integration.ts                                      
- basic                                 packages/models/vite.config.unit.ts                                             
- basic                                 packages/nx-plugin/vite.config.integration.ts                                   
- basic                                 packages/nx-plugin/vite.config.unit.ts                                          
- basic                                 packages/plugin-coverage/vite.config.integration.ts                             
- basic                                 packages/plugin-coverage/vite.config.unit.ts                                    
- basic                                 packages/plugin-eslint/vite.config.integration.ts                               
- basic                                 packages/plugin-eslint/vite.config.unit.ts                                      
- basic                                 packages/plugin-js-packages/vite.config.integration.ts                          
- basic                                 packages/plugin-js-packages/vite.config.unit.ts                          
- basic                                 packages/plugin-lighthouse/vite.config.integration.ts                           
- basic                                 packages/plugin-lighthouse/vite.config.unit.ts                                  
- basic                                 packages/utils/vite.config.integration.ts                                       
- basic                                 packages/utils/vite.config.unit.ts                                              
- basic                                 testing/test-utils/vite.config.unit.ts                                                                                      

**Unused exports (23)**  
- ✅ renderUploadAutorunHint                 function  packages/cli/src/lib/collect/collect-command.ts:42:17     
can be fixed by removing the export form `duplicateRefsInCategoryMetricsErrorMsg`
```ts
// 👇 fixed by removing this export
export function renderUploadAutorunHint(...) { /*...*/ }
``` 
- ✅ duplicateRefsInCategoryMetricsErrorMsg  function  packages/models/src/lib/category-config.ts:54:17                
- ✅ unrefinedCoreConfigSchema               unknown   packages/models/src/lib/core-config.ts:11:14                    
- ✅ refineCoreConfig                        function  packages/models/src/lib/core-config.ts:32:17                    
- ✅ messageStyles                           unknown   packages/utils/src/lib/progress.ts:13:14  
- ✅ singletonUiInstance                     unknown   packages/utils/src/lib/logging.ts:23:12                         
- ✅ logListItem                             function  packages/utils/src/lib/logging.ts:39:17     
- ✅ onlyPluginsOption                       unknown   packages/cli/src/lib/implementation/only-plugins.options.ts:4:14                                          
- ✅ h                                       function  packages/utils/src/lib/reports/md/headline.ts:13:17                   
can be fixed by exporting `h` in `index.ts`
```ts
// 👇 fixed by adding this export
export * from './src/lib/reports/md'
``` 
- ✅ h4                                      function  packages/utils/src/lib/reports/md/headline.ts:29:17             
- ✅ h5                                      function  packages/utils/src/lib/reports/md/headline.ts:33:17             
- ✅ h6                                      function  packages/utils/src/lib/reports/md/headline.ts:37:17 
can be fixed by exporting `groupMetaSchema` in `index.ts`
```ts
// 👇 fixed by adding this export
export {groupMetaSchema} from '.packages/models/src/lib/audit.ts'
```      
Similar fix can be applied for the following issues.
- ✅ throwIsNotPresentError                  function  packages/utils/src/lib/reports/utils.ts:273:17                  
- ✅ groupMetaSchema                         unknown   packages/models/src/lib/group.ts:20:14 
- ✅ NoExportError                           class     packages/utils/src/lib/file-system.ts:78:14                                              
- ✅ weightSchema                            unknown   packages/models/src/lib/implementation/schemas.ts:140:14        
- ✅ descriptionSchema                       unknown   packages/models/src/lib/implementation/schemas.ts:40:14         
- ✅ docsUrlSchema                           unknown   packages/models/src/lib/implementation/schemas.ts:49:14         
- ✅ pluginDataSchema                        unknown   packages/models/src/lib/plugin-config.ts:30:14                  
- ✅ coverageTypeSchema                      unknown   packages/plugin-coverage/src/lib/config.ts:3:14                 
- ✅ coverageResultSchema                    unknown   packages/plugin-coverage/src/lib/config.ts:6:14                 
- ⏳ WORKDIR                                 unknown   packages/plugin-eslint/src/lib/runner/index.ts:14:14            
- ⏳ shieldsBadge                            function  packages/utils/src/lib/reports/utils.ts:71:17

**Unused exported types (12)**  
- ✅ FontStyle              type  packages/utils/src/lib/reports/md/font-style.ts:8:13              
can be fixed by exporting `h` in `index.ts`
```ts
// 👇 fixed by adding this export
export * from './src/lib/reports/md'
```
Same is applicable for the following issues 
- ✅ Hierarchy              type  packages/utils/src/lib/reports/md/headline.ts:2:13                
- ✅ Order                  type  packages/utils/src/lib/reports/md/list.ts:1:13
- ✅ GroupMeta              type  packages/models/src/lib/group.ts:26:13                            
can be fixed by exporting `GroupMeta` in `index.ts`
```ts
// 👇 fixed by adding this export
export {GroupMeta} from '.packages/models/src/lib/group.ts'
```     
- ✅ CliUiBase              type  packages/utils/src/lib/logging.ts:8:13     
can be fixed by removing the export form `CliUiBase`
```ts
// 👇 fixed by removing this export
export type CliUiBase = /*...*/;
``` 
- ✅ OnlyPluginsCliOptions  type  packages/cli/src/lib/implementation/only-plugins.model.ts:4:13    
this type is not used. 
It can be fixed by using it to ensure the option names.
```ts
export function yargsOnlyPluginsOptionsDefinition(): Record<
   // 👇 fixed by replacing 'onlyPlugins' with `keyof OnlyPluginsCliOptions`
  'onlyPlugins',
  Options
> {
  return {
    onlyPlugins: onlyPluginsOption,
  };
}
```

- ✅ HistoryOnlyOptions     type  packages/core/src/lib/history.ts:8:13    
is unused can be fixed by exporting it. (lands in #273)                         
- ✅ LighthouseCliOptions   type  packages/plugin-lighthouse/src/lib/utils.ts:16:13 
is unused can be fixed by exporting it. (lands in #549)          
- ✅  Yarnv2AuditAdvisory    type  packages/plugin-js-packages/src/lib/runner/audit/types.ts:87:13   
can be fixed by exporting `Yarnv2AuditAdvisory` in `index.ts`
```ts
// 👇 fixed by adding this export
export {Yarnv2AuditAdvisory} from './src/lib/runner/audit/types.ts'
```     
Same applies for the following exported types
- ✅ NpmVersionOverview     type  packages/plugin-js-packages/src/lib/runner/outdated/types.ts:19:13
- ✅ Yarnv1VersionOverview  type  packages/plugin-js-packages/src/lib/runner/outdated/types.ts:33:13
- ✅ Yarnv2VersionOverview  type  packages/plugin-js-packages/src/lib/runner/outdated/types.ts:54:13                 
                       

**Duplicate exports (2)**  
- ✅ addToProjectGenerator, default  packages/nx-plugin/src/generators/configuration/generator.ts
the audit can be fixed by only exporting the same instance 1 time. Note: also applies for the following autit.
```ts
// packages/nx-plugin/src/generators/configuration/generator.ts
// 👇 e.g. fixed by removing this export and keep the default export
export function initGenerator(tree: Tree, schema: InitGeneratorSchema) {
   // ...
}
// 👇 e.g. fixed by removing this export and keep the explicite export
export default initGenerator;
```
- ✅ initGenerator, default          packages/nx-plugin/src/generators/init/generator.ts

### Snippet to test fixes:

`npm remove cli-table3 conventional-changelog-angular eslint-plugin-jsx-a11y tsx && npm i -D jsonc-eslint-parser lighthouse-logger debug chrome-launcher minimatch`

## Screenshots of the implementation in different views

|  Plugin Terminal Output | Plugin MD Output |
| --- | --- | 
| Screenshot 2024-03-14 at 09 02 59 |  Screenshot 2024-03-15 at 12 23 04 Screenshot 2024-03-14 at 11 29 39 | 

|  Portal | Portal |
| --- | --- |
|  Screenshot 2024-03-16 at 00 47 23 | Screenshot 2024-03-16 at 00 49 20 | 

Related:
https://github.com/webpro/knip/issues/551

### CODE REVIEW

File: examples/plugins/src/knip/README.md [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+
+🕵️ **Code PushUp plugin for detecting unused code** 📋
+
+---
+
+## Getting started
+
+1. If you haven't already, install [@code-pushup/cli](../cli/README.md) and create a configuration file.
+
+2. Copy the [plugin source](../file-size) as is into your project
```
Comment (beaussan) [inline examples/plugins/src/knip/README.md:11]: **issue:** Wrong name here‏

File: examples/plugins/src/knip/README.md [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+     // ...
+     categories: [
+       // ...
+       {
+         slug: 'performance',
+         title: 'Performance',
+         refs: [...fileSizeRecommendedRefs],
+       },
+     ],
+   };
```
Comment (beaussan) [inline examples/plugins/src/knip/README.md:48]: **issue:** ‏Wrong plugin here

File: .github/workflows/ci.yml [hunk 3]
```diff
@@ -105,6 +105,16 @@ jobs:
...
+      - name: Set custom Chrome path for Windows only
+        if: matrix.os == 'windows-latest'
+        # This path is considered in `testing/setup/src/lib/chrome-path-setup.ts` and used in different test configurations
+        run: |
+          echo "CUSTOM_CHROME_PATH=C:\\Program Files\\Google\\Chrome\\Application\\chrome.exe" | Out-File -FilePath $env:GITHUB_ENV -Append -Encoding utf8
+        shell: pwsh
+      - name: enable debugging for Lighthouse
+      #- name: Log all environment variables
+      #  run: |
+      #      printenv
```
Comment (BioPhoton) [inline .github/workflows/ci.yml:117]:
```suggestion
```

File: CONTRIBUTING.md [hunk 4]
```diff
@@ -12,6 +12,25 @@ Make sure to install dependencies:
...
+
+**❗️** Test Inclusion Logic
+
+- `INCLUDE_SLOW_TESTS='false'` skips long tests.
+- without `INCLUDE_SLOW_TESTS`, tests run if `CI` is set.
+
+**❗️❗️** Windows specific path set only in CI
+
+- some setups also require this setting locally
+
```
Comment (BioPhoton) [inline CONTRIBUTING.md:33]:
```suggestion
```

**Comment by matejchalk** [comment]
The _Unused development dependencies_ audit is full of false positives, I'm afraid :disappointed: It doesn't seem to detect usage other than `import`/`require` in code. Other than the false positives already mentioned, the following are also not valid:

> ✅ tsx package.json
> This package is not used anywhere in the code either directly on in "scripts" or configuration files.
> You can fix it by running `npm remove tsx`
> Same goes for the follwing dependencies
> ✅ @trivago/prettier-plugin-sort-imports package.json

- [`@trivago/prettier-plugin-sort-imports`](https://github.com/trivago/prettier-plugin-sort-imports) is used, Prettier picks it up automatically, `importOrder` and `importOrderSortSpecifiers` options configured in [`.prettierrc`](https://github.com/code-pushup/cli/blob/main/.prettierrc#L5-L10)
  - could also be made explicit by adding `plugins: ['@trivago/prettier-plugin-sort-imports']` 
- `tsx` is used by the `perf` Nx target in [`packages/utils/project.json`](https://github.com/code-pushup/cli/blob/main/packages/utils/project.json#L33)

**Comment by beaussan** [comment]
> ❓ eslint-plugin-react package.json
> should have been picked up by the nx helper WDYT @beaussan
> the following packages are (most probably) a cascade of this issue
> ❓ eslint-plugin-react-hooks package.json

This was the case since we excluded the example projects, I changed that in the base config and this is now properly reported


> ❌ commitlint-plugin-tense package.json
> false as it is used in commitlint.config.js

This is something missing on the knip side, it will only look at the extends but not the plugins array. I will make a follow up on their side to update it. In the meantime, added to the ignore array

> ❓ basic e2e/cli-e2e/vite.config.e2e.ts
> This shows no dependents when running npm ls basic
> This could be fixed by installing basic manually over npm i basic -D but seems odd as there are no deps listed.
> Same goes for all other issues related to basic

This is a false positive, it shouldn't require a npm lib since it's a built in of vitest. I will make a pr to add it to the [built in reporters on the knip side](https://github.com/webpro/knip/blob/main/packages/knip/src/plugins/vitest/helpers.ts#L29) so this shouldn't occur anymore. In the meantime, I added it to the ignore dependencies array


> tsx is used by the perf Nx target in [packages/utils/project.json](https://github.com/code-pushup/cli/blob/main/packages/utils/project.json?rgh-link-date=2024-04-02T09%3A34%3A45Z#L33)

This should have been picked up by knip, since it will look at the project json configs to detect the scripts and thrid party requirements, I will follow up with an issue on their side

**Comment by beaussan** [comment]
Related PRs :
https://github.com/webpro/knip/pull/587
https://github.com/webpro/knip/pull/588

**Pull request closed without merging**